dotnet/src/webdriver/Firefox/Preferences.cs

```diff
@@ -152,39 +156,47 @@ private static bool IsWrappedAsString(string value)
         {
-            return !this.immutablePreferences.ContainsKey(preferenceName);
+            return !this.immutablePreferences.Contains(preferenceName);
         }
 
-        private void SetPreferenceValue(string key, object value)
+        private void SetPreferenceValue(string key, JsonNode? value)
         {
             if (!this.IsSettablePreference(key))
             {
-                string message = string.Format(CultureInfo.InvariantCulture, "Preference {0} may not be overridden: frozen value={1}, requested value={2}", key, this.immutablePreferences[key], value.ToString());
+                string message = string.Format(CultureInfo.InvariantCulture, "Preference {0} may not be overridden: frozen value={1}, requested value={2}", key, this.preferences[key], value?.ToString());
                 throw new ArgumentException(message);
             }
 
-            string stringValue = value as string;
-            if (stringValue != null)
+            if (value is JsonValue jValue)
             {
-                if (IsWrappedAsString(stringValue))
+                if (jValue.TryGetValue(out string? stringValue))
                 {
-                    throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, "Preference values must be plain strings: {0}: {1}", key, value));
+                    if (IsWrappedAsString(stringValue))
+                    {
+                        throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, "Preference values must be plain strings: {0}: {1}", key, value));
+                    }
+
+                    this.preferences[key] = string.Format(CultureInfo.InvariantCulture, "\"{0}\"", value);
+                    return;
                 }
 
-                this.preferences[key] = string.Format(CultureInfo.InvariantCulture, "\"{0}\"", value);
-                return;
-            }
+                if (jValue.TryGetValue(out bool boolValue))
+                {
+                    this.preferences[key] = boolValue ? "true" : "false";
+                    return;
+                }
 
-            if (value is bool)
-            {
-                this.preferences[key] = Convert.ToBoolean(value, CultureInfo.InvariantCulture).ToString().ToLowerInvariant();
-                return;
-            }
+                if (jValue.TryGetValue(out int intValue))
+                {
+                    this.preferences[key] = intValue.ToString(CultureInfo.InvariantCulture);
+                    return;
+                }
 
-            if (value is int || value is long)
-            {
-                this.preferences[key] = Convert.ToInt32(value, CultureInfo.InvariantCulture).ToString(CultureInfo.InvariantCulture);
-                return;
+                if (jValue.TryGetValue(out long longValue))
+                {
+                    this.preferences[key] = ((int)longValue).ToString(CultureInfo.InvariantCulture);
```

Comment:
This is just to maintain the same exception-for-exception handling as before. Let me know if JSON numbers > `int.MaxValue` should be handled differently.